### PR TITLE
QueueIterators for detached sinks are dropped when sources have stopped

### LINKED ISSUE #87: Memory leak
It looks like `Sink::detach` will leak the associated `QueueIterator` (even when no sounds are, or can be played anymore) until the stream for the associated endpoint is dropped.

This means if you create and detach too many sinks it degrades sound quality.

### PR DESCRIPTION
**Pull request by Eroc33**
Fixes #87
Edit: Not sure why but this seems to be preventing the first sound on an endpoint from playing, so it's probably not ready to be merged yet.

### CODE REVIEW

File: src/engine.rs [hunk 1]
```diff
@@ -104,18 +104,27 @@ impl Engine {
...
                 {
                     let mut pending = epv.pending_sounds.lock().unwrap();
                     sounds.append(&mut pending);
                 }
 
+
                 if sounds.len() == 0 {
                     return Ok(());
                 }
                 sounds.retain(|s| !s.1.local_dead);
```
Comment (Xaeroxe23) [inline src/engine.rs]: I'd recommend merging these two calls to retain, we can take advantage of some short circuiting optimization if we do.

File: src/engine.rs [hunk 2]
```diff
@@ -104,18 +104,27 @@ impl Engine {
...
+                //Drop if it's not playing a real source, and it's sink is detached
+                sounds.retain(|s|{
+                    if !s.1.is_playing_real_source{
+                        Arc::strong_count(&s.1.next)>1
```
Comment (Xaeroxe23) [inline src/engine.rs]: [strong_count](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.strong_count) is considered unstable.  We should look for some alternate implementation.

Comment (Xaeroxe23) [inline src/engine.rs]: Also, that logic needed to be inverted.  This was the cause of the issue you mentioned:

Edit: Not sure why but this seems to be preventing the first sound on an endpoint from playing, so it's probably not ready to be merged yet.

**Comment by Xaeroxe23** [comment]
https://github.com/tomaka/rodio/pull/89 This is my alternative implementation.  I like this because it doesn't use strong_count which is unstable.

**Comment by Xaeroxe23** [comment]
I'm also not sure why this version was experiencing the problem with not playing the first sound.  I tried debugging that myself but #89 doesn't seem to suffer from that issue thankfully.

**Comment by Eroc33** [comment]
Closing in favour of #89

**Pull request closed without merging**